ComponentKit/Core/CKComponentMemoizer.mm

```diff
@@ -0,0 +1,189 @@
+  _CKComponentMemoizerImpl *_next;
+
+  // maps CKMemoizationKey -> any number of CKComponent *
+  std::unordered_multimap<CKMemoizationKey, CKComponent *> componentCache_;
+
+  std::unordered_map<CKLayoutMemoizationKey, CKComponentLayout, CKLayoutMemoizationKey::Hash, CKLayoutMemoizationKey::Equals> layoutCache_;
+}
+
+@end
+
+
+@implementation _CKComponentMemoizerImpl
+
+- (instancetype)init
+{
+  self = [super init];
+  if (!self) return nil;
+
+  return self;
+}
+
+- (void)dealloc
+{
+  
+}
+
+- (CKComponent *)dequeueComponentForKey:(CKMemoizationKey)key
+{
+  auto it = componentCache_.find(key);
+  if (it != componentCache_.end()) {
+    CKComponent *c = it->second;
+    // Remove this component from the cache, since you can't mount a component twice
+    componentCache_.erase(it);
+    return c;
+  }
+  return nil;
+}
+
+- (_CKComponentMemoizerImpl *)next
+{
+  if (!_next) {
+    _next = [[_CKComponentMemoizerImpl alloc] init];
+  }
+  return _next;
+}
+
+- (void)enqueueComponent:(CKComponent *)component forKey:(CKMemoizationKey)key
+{
+  self.next->componentCache_.insert({key, component});
+}
+
+- (CKComponentLayout)cachedLayout:(CKComponent *)component thatFits:(CKSizeRange)constrainedSize restrictedToSize:(CKComponentSize)size parentSize:(CGSize)parentSize
+{
+  CKLayoutMemoizationKey key{.component = component, .thatFits = constrainedSize, .parentSize = parentSize};
+  auto it = layoutCache_.find(key);
+  if (it != layoutCache_.end()) {
+    return it->second;
+  } else {
+    CKComponentLayout layout = [component computeLayoutThatFits:constrainedSize restrictedToSize:size relativeToParentSize:parentSize];
+    self.next->layoutCache_.insert({key, layout});
+    return layout;
+  }
+}
+
++ (_CKComponentMemoizerImpl *)currentMemoizer
+{
+  return [[NSThread currentThread] threadDictionary][CKComponentMemoizerThreadKey];
+}
+
++ (void)setCurrentMemoizer:(_CKComponentMemoizerImpl *)memoizer
+{
+  if (memoizer) {
+    [[NSThread currentThread] threadDictionary][CKComponentMemoizerThreadKey] = memoizer;
+  } else {
+    [[[NSThread currentThread] threadDictionary] removeObjectForKey:CKComponentMemoizerThreadKey];
+  }
+}
+
+@end
+
+CKComponentMemoizer::CKComponentMemoizer(id previousMemoizerState)
+{
+  _CKComponentMemoizerImpl *mipl = previousMemoizerState ?: [[_CKComponentMemoizerImpl alloc] init];
+
+  // Push this memoizer onto the current thread
+  id current = [_CKComponentMemoizerImpl currentMemoizer];
+  previousMemoizer_ = current;
+  [_CKComponentMemoizerImpl setCurrentMemoizer:mipl];
+};
+
+CKComponentMemoizer::~CKComponentMemoizer()
+{
+  // Pop memoizer
+  [_CKComponentMemoizerImpl setCurrentMemoizer:previousMemoizer_];
+}
+
+id CKMemoize(CKMemoizationKey memoizationKey, id (^block)(void))
+{
+  _CKComponentMemoizerImpl *impl = [_CKComponentMemoizerImpl currentMemoizer];
+  CKComponent *component = [impl dequeueComponentForKey:memoizationKey];
+  if (!component) {
+    component = block();
```

Comment:
It's probably a bug either way, but mind adding a `NULL`-check lest we dereference `NULL` and `SEGFAULT`.